src/classes/CON_ContactMerge_CTRL.cls

```diff
@@ -596,30 +614,51 @@ public with sharing class CON_ContactMerge_CTRL {
         }
         return false;
     }
-    
+
+     /*******************************************************************************************************
+    * @description checks if the logged in user have delete permission for the selected records.
+    * @param ContactIds list of contacts
+    * @return boolean.  true if delete access is allowed.
+    ********************************************************************************************************/
+    private boolean isUserHasDeleteAccess(set<Id> ContactIds) {
+        //This is the logic which checks if delete permission for the selected records exists or not
+        List <UserRecordAccess> userAccess = [
+             SELECT RecordId, HasEditAccess, HasDeleteAccess, MaxAccessLevel
+             FROM UserRecordAccess
+             WHERE UserId = :UserInfo.getUserId()
+             And RecordId IN :ContactIds
+        ];
+        for (UserRecordAccess user: userAccess) {
+            if (!user.HasDeleteAccess && user.MaxAccessLevel != 'None') {
```

Comment:
There is a scenario in the code where it checks if any records gets deleted(deleted by background process or something) after selecting the contact for the contact merge and it provides different error message for that scenario. However, that is getting caught by this method and provides this error msg instead of the other message. The two test methods failed because of that. I noticed that if a record is deleted in the process, maxaccesslevel changes to 'none', all access like editaccess,readaccess etc becomes false.  I added this criteria to resolve the failures. However, there is  an alternative to do another query for the contact whether isdeleted is false or not. But it becomes more complicated in the functionality. I can talk to you and explain about this in detail if needed.

Reply:
The methods in the test class are failNextStepWithDeletedRecordSelectedOne2One and failNextStepWithDeletedRecordSelected.

Reply:
Ah. I see the issue. If the Contact is deleted, we want the error message to reflect that the Contact has been deleted instead of that they don't have delete permissions.

I can't find documentation that explicitly states the behavior of the `MaxAccessLevel` field for deleted records. That makes me a bit nervous about relying on the value of that field for undocumented behavior. Were you able to find something that called this out?

I suppose an alternative approach is to move the call to canUserDeleteSelectedContacts() to after the query, towards the bottom of the method. This way the method can do all the logic that queries the records first. It might make the method more readable if the entire block of code that builds the `fieldRows` list (effectively from line 284 to 431) is abstracted out into a method (something like `buildFieldListForSelectedContacts()`.

Reply:
@force2b , If MaxAccessLevel is none, then, the record don't have even read permission on it. Without that, the user can't even select the contact for merging it. That's why I selected that criteria to bypass the deleted contacts in the logic. I will check if any documentation is there too.
For your alternative approach, I am not sure if I can move the call after as you mentioned. Because, lines 284 to 347, it checks if any contacts got deleted in the process and shows that query error message for that scenario. From lines 348 to 431, it does prework for the contact merge process for step3. We don't want to do that if the contact don't have record level access.

Reply:
@force2b , I have moved the method to line 342 for this check after the query check. I have removed the Maxaccesslevel check in the criteria.